driver-core/src/main/com/mongodb/connection/CommandMessage.java

```diff
@@ -129,15 +133,22 @@ protected EncodingMetadata encodeMessageBodyWithMetadata(final BsonOutput bsonOu
             addDocument(getCommandToEncode(), bsonOutput, commandFieldNameValidator, getExtraElements(sessionContext));
 
             if (payload != null) {
+                int payloadStartPosition = payload.getPosition();
                 bsonOutput.writeByte(1);          // payload type
-                int payloadPosition = bsonOutput.getPosition();
+                int payloadBsonOutputStartPosition = bsonOutput.getPosition();
                 bsonOutput.writeInt32(0);         // size
                 bsonOutput.writeCString(payload.getPayloadName());
                 writePayload(new BsonBinaryWriter(bsonOutput, payloadFieldNameValidator), bsonOutput, getSettings(),
                         messageStartPosition, payload);
 
-                int payloadLength = bsonOutput.getPosition() - payloadPosition;
-                bsonOutput.writeInt32(payloadPosition, payloadLength);
+                int payloadBsonOutputLength = bsonOutput.getPosition() - payloadBsonOutputStartPosition;
+                bsonOutput.writeInt32(payloadBsonOutputStartPosition, payloadBsonOutputLength);
+
+                if (sessionContext.hasActiveTransaction()) {
+                    if (payload.getPosition() > payloadStartPosition + 1) {
```

Comment:
Advances the statement id by n - 1 additional statements, one for each included payload item beyond the first one.